### PR TITLE
feat: update category matching algorithm

### PR DESCRIPTION
**Pull request by raphael0202**
### What
- See #918

### Part of
- #379

### CODE REVIEW

**Review by alexgarel — APPROVED** [review]
Great idea and implementation.

I have some comments, most are style, but some might be important.
However I approve and let you fix what you want to fix.

File: robotoff/utils/text.py [hunk 1]
```diff
@@ -41,13 +38,19 @@ def strip_consecutive_spaces(text: str) -> str:
...
+    return spacy.blank(lang)
+
+
+@functools.lru_cache()
+def get_lemmatizing_nlp(lang: str) -> spacy.Language:
+    """Return a spaCy language pipeline with a lookup lemmatizer."""
+    nlp = spacy.blank(lang)
+    nlp.add_pipe("lemmatizer", config={"mode": "lookup"})
+    nlp.initialize()
+    return nlp
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/utils/text.py:53]: I used `@functools.lru_cache()` instead. I want to move away from the `CacheStore` class to use caching libraries (either cachetools or built-in functools.lru_cache), which is less verbose to use (and probably more optimized).

File: robotoff/prediction/category/matcher.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    ),
+    "en": re.compile(r"organic|pgi"),
+    "es": re.compile(r"ecológico|ecológica|extra|comercio justo"),
+    "de": re.compile(r"bio"),
+    "nl": re.compile(r"bio"),
+    "it": re.compile(
+        r"biologica|biologico|commercio solidale|commercio equo e solidale"
+        r"|da agricoltura biologica"
+    ),
+}
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:40]: I've always had mixed feelings about importing this kind of things from the taxonomies directly. On the pros side, it makes maintenance easier. On the cons side, we've tested the algorithm with specific processing, and if the processing changes automatically, new types of errors may be introduced silently.

File: robotoff/app/api.py [hunk 3]
```diff
@@ -403,28 +407,55 @@ def on_get(self, req: falcon.Request, resp: falcon.Response):
...
+                    {
+                        "value_tag": p.value_tag,
+                        "lang": p.data["lang"],
+                        "product_name": p.data["product_name"],
+                        "pattern": p.data["pattern"],
+                    }
+                    for p in itertools.chain.from_iterable(
+                        matcher_predict_by_lang(product).values()
+                    )
+                ]
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/app/api.py:456]: Indeed it's better this way, I'm creating a new `predict_category` function in `robotoff/prediction/category/__init__.py`

File: robotoff/app/api.py [hunk 4]
```diff
@@ -403,28 +407,55 @@ def on_get(self, req: falcon.Request, resp: falcon.Response):
...
-                for p in predictions
-            ]
-
-        resp.media = {"neural": categories}
+            if "neural" in predictors:
+                deepest_only: bool = req.media.get("deepest_only", False)
+                threshold: Optional[float] = req.media.get("threshold")
+                predictions = CategoryClassifier(
+                    get_taxonomy(TaxonomyType.category.name)
+                ).predict(product, deepest_only, threshold)
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/app/api.py:434]: Having predicted that are parent of each other almost never happens for matcher predictor (it means we predict cat1 for lang1 and cat2 for lang2 and cat1 is a parent of cat2 or the other way around). But for clarity and consistency it makes sense to filter matcher results too :+1:

File: robotoff/prediction/category/matcher.py [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+            )
+            for name in names:
+                processed = process(name, lang)
+                # Don't add it to the store if empty or if it's in the
+                # exception list
+                if processed and processed not in match_exceptions:
+                    if processed not in store[lang]:
+                        store[lang][processed] = []
+                    if node.id not in (item[0] for item in store[lang][processed]):
+                        store[lang][processed].append((node.id, name))
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:172]: I think a KeyError may be raised as `store[lang][processed]` may not exist during condition check

File: robotoff/prediction/category/matcher.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+    (value_tag, original_name) tuples. Queries are processed through
+    the `process` function. We have a match if the processed query is a key of
+    the map. It provides a very fast (although not memory-efficient) way of
+    performing matching.
+
+    :param taxonomy_type: the taxonomy to build the match maps from
+    :return: the match maps
+    """
+    taxonomy = get_taxonomy(taxonomy_type)
+    store: Dict[str, Dict[str, List[Tuple[str, str]]]] = {}
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:150]: What I don't like about defaultdict is that they don't render good:
```python
In [4]: x = defaultdict(lambda: defaultdict(list))

In [5]: x
Out[5]: defaultdict(<function __main__.<lambda>()>, {})

In [6]: x["3"]["2"].append(4)

In [7]: x
Out[7]: 
defaultdict(<function __main__.<lambda>()>,
            {'3': defaultdict(list, {'2': [4]})})
```

File: robotoff/prediction/category/matcher.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+
+def match(query: str, lang: str) -> List[Tuple[str, str, str, str, Tuple[int, int]]]:
+    """Perform category matching.
+
+    If a match is also an ingredient (see
+    `generate_intersect_categories_ingredients`), we only keep the match if
+    it starts at the beginning of the query. It prevents many over-matchings
+    with ingredient names in the query that are not the product category.
+
+    :param query: the unprocessed category name
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:251]: Product name indeed :+1:

File: robotoff/prediction/category/matcher.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+                    "pattern": match[2],
+                    "processed_product_name": match[3],
+                    "start_idx": match[4][0],
+                    "end_idx": match[4][1],
+                    "is_full_match": (match[4][1] - match[4][0] == len(match[3])),
+                },
+                automatic_processing=False,
+                predictor="matcher",
+            )
+            for match in matches
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:316]: indeed way more readable this way!

File: robotoff/prediction/category/matcher.py [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    """
+    return load_json(
+        settings.CATEGORY_MATCHER_MATCH_MAPS[taxonomy_type], compressed=True
+    )
+
+
+@cachetools.cached(cache=cachetools.TTLCache(maxsize=1, ttl=3600))
+def get_processors() -> Dict[str, KeywordProcessor]:
+    """Return a dict mapping lang to KeywordProcessor used to perform category
+    matching."""
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:191]: I always follow flake8 convention to keep comments and docstrings <= 79 chars. So the docstrings are the longest they can be while still be <80 chars.

File: robotoff/prediction/category/matcher.py [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+                if processed and processed not in match_exceptions:
+                    if processed not in store[lang]:
+                        store[lang][processed] = []
+                    if node.id not in (item[0] for item in store[lang][processed]):
+                        store[lang][processed].append((node.id, name))
+
+    return store
+
+
+@cachetools.cached(cache=cachetools.Cache(maxsize=2))
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:177]: probably not, indeed!

File: robotoff/prediction/category/matcher.py [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    the query. Therefore, if more than one category is predicted, it means
+    there was two non-overlapping matches: we don't return any prediction for
+    that lang in that case.
+
+    :param product: product properties
+    :return: a list of category Prediction"""
+    predictions_by_lang = predict_by_lang(product)
+    predictions = []
+    for _, lang_predictions in predictions_by_lang.items():
+        if len(lang_predictions) != 1:
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply robotoff/prediction/category/matcher.py:334]: It's quite unlikely, as flashtext always selects the longuest match: it means that two different patterns for the same category name (and same language) matched the query at different positions. But you're right I should add check this!

File: robotoff/prediction/category/matcher.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply robotoff/prediction/category/matcher.py]: yes, I understand, but the practice I advocate is to cut at meaningful point (but if see the point, that was just to share)

**Pull request merged**